### DISCUSSION TITLE
Is there a complete example for uploading big and small files in Drive?

### DISCUSSION THREAD

**Discussion by piergiuseppepezzoli** [microsoftgraph/msgraph-sdk-python#967] [Q&A]
Right now there are many questions asking for clarification about uploading small and large files to the Drive.
Many answers, however, only give a piece of information or remand to pages of the documentation which are not clear.

For example, https://github.com/microsoftgraph/msgraph-sdk-python/issues/869 gives a nice answer about the uploading problem, BUT uses functions without specifying from where to import them (e.g. DriveItemUploadableProperties. In other answers this is better explained, like in https://github.com/microsoftgraph/msgraph-sdk-python/issues/894 there is the complete explanation on the import).

Another example of an answer which doesn't give the complete picture is https://github.com/microsoftgraph/msgraph-sdk-python/discussions/874, which remands at the documentation for PUT requests https://learn.microsoft.com/en-us/graph/api/driveitem-createuploadsession?view=graph-rest-1.0, but there there isn't any example shown in python. 

I also find that the languages supported in the documentation are not consistent, sometimes python is present, sometimes it is not.

I think that it would be extremely helpful to have a complete example for uploading large and small files using the SDK, with: 
- all imports shown
- the whole procedure reported, complete with the PUT request or upload request

This would help tremendously beginners like me who keep finding incomplete answers about the topic and cannot manage to solve what I believe is one of the most important functions that the SDK should provide

**Comment by michaelhaar** [comment]
@piergiuseppepezzoli Did you find an example?

**Reply by piergiuseppepezzoli** [reply]
Nope, the documentation is still extremely unclear. I can share some of my code to upload files, but I don't think that is very good. I hope that it still helps...

```
from azure.identity.aio import ClientSecretCredential
from msgraph import GraphServiceClient
from msgraph.generated.drives.item.items.item.create_upload_session.create_upload_session_post_request_body import (
    CreateUploadSessionPostRequestBody,
)
from msgraph.generated.models.drive_item_uploadable_properties import DriveItemUploadableProperties

msgraph_settings = {
    "tenat_id": "your_tenant_id",
    "client_id": "your_client_id",
    "client_secret": "your_client_secret",
    "shrd_id": "your_shrd_id",
}

credential = ClientSecretCredential(
    tenant_id=msgraph_settings.tenant_id,
    client_id=msgraph_settings.client_id,
    client_secret=msgraph_settings.client_secret,
)

client = GraphServiceClient(
    credentials=credential,
    scopes=[
        "https://graph.microsoft.com/.default",
    ],
)


async def upload_small_file(destination_path: str, file_content):
    uploadable_properties = DriveItemUploadableProperties(
        additional_data={"@microsoft.graph.conflictBehavior": "replace"}
    )
    upload_session_request_body = CreateUploadSessionPostRequestBody(item=uploadable_properties)
    # can be used for normal drive uploads
    try:
        upload_session = (
            await client.drives.by_drive_id(msgraph_settings.shrd_id)
            .items.by_drive_item_id(f"root:/{destination_path}:")
            .create_upload_session.post(upload_session_request_body)
        )
        if upload_session:
            print(f"Created upload session for {destination_path}")

        uploaded_file = (
            await client.drives.by_drive_id(msgraph_settings.shrd_id)
            .items.by_drive_item_id(f"root:/{destination_path}:")
            .content.put(file_content)
        )
        if uploaded_file:
            print(f"Uploaded file: {destination_path}")
    except Exception as e:
        print(f"Error creating upload session: {e}")
        raise e
```
